### DISCUSSION TITLE
On how to load some prior knowledge of known formulas

### DISCUSSION THREAD

**Discussion by Eachen-Soong** [MilesCranmer/PySR#378] [Q&A]
Hi, Miles!

I'm recently testing your excellent tool on explainable AI, so I wish I could somehow merge the PySR workflow better with neural networks.
On the official website of PySR, I only saw that you can load 'pretrained models' from a pickle file. 
But is it feasible to directly load a prior formula with sympy?
Say, you've known the classic momentum formula: p = m * v, can you input this prior knowledge, and let PySR to regress out the one in special relativity?

Much appreciated!

**Comment by MilesCranmer** [comment]
Hi @NoRatRacing,

Great question. Unfortunately there isn’t anything built in at the moment. It shouldn’t be too bad to add it. If you have some time available I could walk you through adding it. I can try myself but have a list of other features I’m working on so it will take a bit longer before I can work on it.

Note that you can restart from where the previous search left off when you use the `warm_start` parameter and call `fit` twice. So I think this feature could exploit that same functionality, and basically pass the prior knowledge as existing equations/population members.

What do you think?

Cheers,
Miles

**Reply by Eachen-Soong** [reply]
Sure, I have time to try adding this feature, and it's necessary for my current researches.
I'm now starting with loading some written .csv files (with my own written formula) with PySRRegressor.from_file(), but it seems I didn't do it right way...
```python

model_loaded = PySRRegressor(
    niterations=1000,
    binary_operators=["plus", "mult", "/"],
    unary_operators=["neg"],
    complexity_of_constants=1000,
    **default_pysr_params
)
model_loaded = model_loaded.from_file('./formula.csv')
```
This code causes an assertion error: assert binary_operators is not None
which means I didn't successfully load the binary_operators...?

**Reply by MilesCranmer** [reply]
When loading a past result, you should load from the `pkl` file. If loading from the csv file, you need to provide the other information such as operators.

Also, for serialization, see https://github.com/MilesCranmer/PySR/discussions/313 for how to load the Julia state (which contains the state information to do a warm start from a file).
